File: snippets/java/src/main/java/develop/TracingMicrometer.java
package develop;

// <start_here>
import dev.restate.sdk.HandlerRunner;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.interceptor.micrometer.MicrometerInterceptorFactory;
import io.micrometer.observation.ObservationRegistry;

class MyMicrometerApp {
  public static void main(String[] args) {
    ObservationRegistry registry = ObservationRegistry.create(); // your ObservationRegistry
    var micrometer = new MicrometerInterceptorFactory(registry);

    var endpoint =
        Endpoint.bind(
                new MyService(),
                new HandlerRunner.Options()
                    .addHandlerInterceptorFactory(micrometer)
                    .addRunInterceptorFactory(micrometer))
            .build();
  }
}
// <end_here>

File: snippets/kotlin/src/main/kotlin/develop/TracingMicrometer.kt
package develop

// <start_here>
import dev.restate.sdk.interceptor.micrometer.kotlin.MicrometerInterceptorFactory
import dev.restate.sdk.kotlin.HandlerRunner
import dev.restate.sdk.kotlin.endpoint.endpoint
import io.micrometer.observation.ObservationRegistry

fun micrometerTracedEndpoint() {
  val registry: ObservationRegistry = ObservationRegistry.create() // your ObservationRegistry
  val micrometer = MicrometerInterceptorFactory(registry)

  val endpoint = endpoint {
    bind(
        MyService(),
        HandlerRunner.Options(
            handlerInterceptorFactories = mutableListOf(micrometer),
            runInterceptorFactories = mutableListOf(micrometer),
        ),
    )
  }
}
// <end_here>
